apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/json/MapJsonReader.kt

```diff
@@ -277,15 +271,16 @@ class MapJsonReader(val root: Map<String, Any?>) : JsonReader {
         skipValue()
       }
     }
+
     return -1
   }
 
   /**
    * Rewinds to the beginning of the current object.
    */
   override fun rewind() {
-    currentIndex = 0
-    currentName = null
+    indexStack[stackSize - 1] = 0
+    nameIndexStack[stackSize - 1] = 0
```

Comment:
This was the actual bug that `nameIndexStack` was not reset on `rewind()`. Everything else is refactoring and small-ish optimizations